File: Microsoft.CodeAnalysis.VisualBasic/Microsoft.CodeAnalysis.VisualBasic.Syntax/TypeStatementSyntax.cs
using System;
using System.ComponentModel;
using Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax;

namespace Microsoft.CodeAnalysis.VisualBasic.Syntax
{
	public abstract class TypeStatementSyntax : DeclarationStatementSyntax
	{
		internal SyntaxNode _attributeLists;

		internal TypeParameterListSyntax _typeParameterList;

		public SyntaxList<AttributeListSyntax> AttributeLists => GetAttributeListsCore();

		public SyntaxTokenList Modifiers => GetModifiersCore();

		public SyntaxToken Identifier => GetIdentifierCore();

		public TypeParameterListSyntax TypeParameterList => GetTypeParameterListCore();

		public int Arity
		{
			get
			{
				if (TypeParameterList != null)
				{
					return TypeParameterList.Parameters.Count;
				}
				return 0;
			}
		}

		public abstract SyntaxToken DeclarationKeyword { get; }

		[EditorBrowsable(EditorBrowsableState.Never)]
		[Obsolete("This member is obsolete. Use DeclarationKeyword or a more specific property (e.g. ClassKeyword) instead.", true)]
		public SyntaxToken Keyword => DeclarationKeyword;

		internal TypeStatementSyntax(GreenNode green, SyntaxNode parent, int startLocation)
			: base(green, parent, startLocation)
		{
		}

		internal virtual SyntaxList<AttributeListSyntax> GetAttributeListsCore()
		{
			SyntaxNode redAtZero = GetRedAtZero(ref _attributeLists);
			return new SyntaxList<AttributeListSyntax>(redAtZero);
		}

		public TypeStatementSyntax WithAttributeLists(SyntaxList<AttributeListSyntax> attributeLists)
		{
			return WithAttributeListsCore(attributeLists);
		}

		internal abstract TypeStatementSyntax WithAttributeListsCore(SyntaxList<AttributeListSyntax> attributeLists);

		public TypeStatementSyntax AddAttributeLists(params AttributeListSyntax[] items)
		{
			return AddAttributeListsCore(items);
		}

		internal abstract TypeStatementSyntax AddAttributeListsCore(params AttributeListSyntax[] items);

		internal virtual SyntaxTokenList GetModifiersCore()
		{
			GreenNode modifiers = ((Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.TypeStatementSyntax)base.Green)._modifiers;
			return (modifiers == null) ? default(SyntaxTokenList) : new SyntaxTokenList(this, modifiers, GetChildPosition(1), GetChildIndex(1));
		}

		public TypeStatementSyntax WithModifiers(SyntaxTokenList modifiers)
		{
			return WithModifiersCore(modifiers);
		}

		internal abstract TypeStatementSyntax WithModifiersCore(SyntaxTokenList modifiers);

		public TypeStatementSyntax AddModifiers(params SyntaxToken[] items)
		{
			return AddModifiersCore(items);
		}

		internal abstract TypeStatementSyntax AddModifiersCore(params SyntaxToken[] items);

		internal virtual SyntaxToken GetIdentifierCore()
		{
			return new SyntaxToken(this, ((Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.TypeStatementSyntax)base.Green)._identifier, GetChildPosition(2), GetChildIndex(2));
		}

		public TypeStatementSyntax WithIdentifier(SyntaxToken identifier)
		{
			return WithIdentifierCore(identifier);
		}

		internal abstract TypeStatementSyntax WithIdentifierCore(SyntaxToken identifier);

		internal virtual TypeParameterListSyntax GetTypeParameterListCore()
		{
			return GetRed(ref _typeParameterList, 3);
		}

		public TypeStatementSyntax WithTypeParameterList(TypeParameterListSyntax typeParameterList)
		{
			return WithTypeParameterListCore(typeParameterList);
		}

		internal abstract TypeStatementSyntax WithTypeParameterListCore(TypeParameterListSyntax typeParameterList);

		public TypeStatementSyntax AddTypeParameterListParameters(params TypeParameterSyntax[] items)
		{
			return AddTypeParameterListParametersCore(items);
		}

		internal abstract TypeStatementSyntax AddTypeParameterListParametersCore(params TypeParameterSyntax[] items);

		public abstract TypeStatementSyntax WithDeclarationKeyword(SyntaxToken keyword);

		[EditorBrowsable(EditorBrowsableState.Never)]
		[Obsolete("This member is obsolete. Use DeclarationKeyword or a more specific property (e.g. WithClassKeyword) instead.", true)]
		public TypeStatementSyntax WithKeyword(SyntaxToken keyword)
		{
			return WithDeclarationKeyword(keyword);
		}
	}
}

File: Microsoft.CodeAnalysis.VisualBasic/Microsoft.CodeAnalysis/VisualBasic/Syntax/TypeStatementSyntax.vb
Imports Microsoft.CodeAnalysis
Imports Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax
Imports System
Imports System.ComponentModel

Namespace Microsoft.CodeAnalysis.VisualBasic.Syntax
	Public MustInherit Class TypeStatementSyntax
		Inherits Microsoft.CodeAnalysis.VisualBasic.Syntax.DeclarationStatementSyntax
		Friend _attributeLists As SyntaxNode

		Friend _typeParameterList As Microsoft.CodeAnalysis.VisualBasic.Syntax.TypeParameterListSyntax

		Public ReadOnly Property Arity As Integer
			Get
				If (Me.TypeParameterList Is Nothing) Then
					Return 0
				End If
				Return Me.TypeParameterList.Parameters.Count
			End Get
		End Property

		Public ReadOnly Property AttributeLists As SyntaxList(Of Microsoft.CodeAnalysis.VisualBasic.Syntax.AttributeListSyntax)
			Get
				Return Me.GetAttributeListsCore()
			End Get
		End Property

		Public MustOverride ReadOnly Property DeclarationKeyword As Microsoft.CodeAnalysis.SyntaxToken

		Public ReadOnly Property Identifier As Microsoft.CodeAnalysis.SyntaxToken
			Get
				Return Me.GetIdentifierCore()
			End Get
		End Property

		<EditorBrowsable(EditorBrowsableState.Never)>
		<Obsolete("This member is obsolete. Use DeclarationKeyword or a more specific property (e.g. ClassKeyword) instead.", True)>
		Public ReadOnly Property Keyword As Microsoft.CodeAnalysis.SyntaxToken
			Get
				Return Me.DeclarationKeyword
			End Get
		End Property

		Public ReadOnly Property Modifiers As SyntaxTokenList
			Get
				Return Me.GetModifiersCore()
			End Get
		End Property

		Public ReadOnly Property TypeParameterList As Microsoft.CodeAnalysis.VisualBasic.Syntax.TypeParameterListSyntax
			Get
				Return Me.GetTypeParameterListCore()
			End Get
		End Property

		Friend Sub New(ByVal green As GreenNode, ByVal parent As SyntaxNode, ByVal startLocation As Integer)
			MyBase.New(green, parent, startLocation)
		End Sub

		Public Function AddAttributeLists(ByVal ParamArray items As Microsoft.CodeAnalysis.VisualBasic.Syntax.AttributeListSyntax()) As Microsoft.CodeAnalysis.VisualBasic.Syntax.TypeStatementSyntax
			Return Me.AddAttributeListsCore(items)
		End Function

		Friend MustOverride Function AddAttributeListsCore(ByVal ParamArray items As Microsoft.CodeAnalysis.VisualBasic.Syntax.AttributeListSyntax()) As Microsoft.CodeAnalysis.VisualBasic.Syntax.TypeStatementSyntax

		Public Function AddModifiers(ByVal ParamArray items As Microsoft.CodeAnalysis.SyntaxToken()) As Microsoft.CodeAnalysis.VisualBasic.Syntax.TypeStatementSyntax
			Return Me.AddModifiersCore(items)
		End Function

		Friend MustOverride Function AddModifiersCore(ByVal ParamArray items As Microsoft.CodeAnalysis.SyntaxToken()) As Microsoft.CodeAnalysis.VisualBasic.Syntax.TypeStatementSyntax

		Public Function AddTypeParameterListParameters(ByVal ParamArray items As Microsoft.CodeAnalysis.VisualBasic.Syntax.TypeParameterSyntax()) As Microsoft.CodeAnalysis.VisualBasic.Syntax.TypeStatementSyntax
			Return Me.AddTypeParameterListParametersCore(items)
		End Function

		Friend MustOverride Function AddTypeParameterListParametersCore(ByVal ParamArray items As Microsoft.CodeAnalysis.VisualBasic.Syntax.TypeParameterSyntax()) As Microsoft.CodeAnalysis.VisualBasic.Syntax.TypeStatementSyntax

		Friend Overridable Function GetAttributeListsCore() As SyntaxList(Of Microsoft.CodeAnalysis.VisualBasic.Syntax.AttributeListSyntax)
			Return New SyntaxList(Of Microsoft.CodeAnalysis.VisualBasic.Syntax.AttributeListSyntax)(MyBase.GetRedAtZero(Me._attributeLists))
		End Function

		Friend Overridable Function GetIdentifierCore() As Microsoft.CodeAnalysis.SyntaxToken
			Return New Microsoft.CodeAnalysis.SyntaxToken(Me, DirectCast(MyBase.Green, Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.TypeStatementSyntax)._identifier, Me.GetChildPosition(2), MyBase.GetChildIndex(2))
		End Function

		Friend Overridable Function GetModifiersCore() As SyntaxTokenList
			Dim syntaxTokenLists As SyntaxTokenList
			Dim green As GreenNode = DirectCast(MyBase.Green, Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.TypeStatementSyntax)._modifiers
			syntaxTokenLists = If(green Is Nothing, New SyntaxTokenList(), New SyntaxTokenList(Me, green, Me.GetChildPosition(1), MyBase.GetChildIndex(1)))
			Return syntaxTokenLists
		End Function

		Friend Overridable Function GetTypeParameterListCore() As Microsoft.CodeAnalysis.VisualBasic.Syntax.TypeParameterListSyntax
			Return MyBase.GetRed(Of Microsoft.CodeAnalysis.VisualBasic.Syntax.TypeParameterListSyntax)(Me._typeParameterList, 3)
		End Function

		Public Function WithAttributeLists(ByVal attributeLists As SyntaxList(Of Microsoft.CodeAnalysis.VisualBasic.Syntax.AttributeListSyntax)) As Microsoft.CodeAnalysis.VisualBasic.Syntax.TypeStatementSyntax
			Return Me.WithAttributeListsCore(attributeLists)
		End Function

		Friend MustOverride Function WithAttributeListsCore(ByVal attributeLists As SyntaxList(Of Microsoft.CodeAnalysis.VisualBasic.Syntax.AttributeListSyntax)) As Microsoft.CodeAnalysis.VisualBasic.Syntax.TypeStatementSyntax

		Public MustOverride Function WithDeclarationKeyword(ByVal keyword As Microsoft.CodeAnalysis.SyntaxToken) As Microsoft.CodeAnalysis.VisualBasic.Syntax.TypeStatementSyntax

		Public Function WithIdentifier(ByVal identifier As Microsoft.CodeAnalysis.SyntaxToken) As Microsoft.CodeAnalysis.VisualBasic.Syntax.TypeStatementSyntax
			Return Me.WithIdentifierCore(identifier)
		End Function

		Friend MustOverride Function WithIdentifierCore(ByVal identifier As Microsoft.CodeAnalysis.SyntaxToken) As Microsoft.CodeAnalysis.VisualBasic.Syntax.TypeStatementSyntax

		<EditorBrowsable(EditorBrowsableState.Never)>
		<Obsolete("This member is obsolete. Use DeclarationKeyword or a more specific property (e.g. WithClassKeyword) instead.", True)>
		Public Function WithKeyword(ByVal keyword As Microsoft.CodeAnalysis.SyntaxToken) As Microsoft.CodeAnalysis.VisualBasic.Syntax.TypeStatementSyntax
			Return Me.WithDeclarationKeyword(keyword)
		End Function

		Public Function WithModifiers(ByVal modifiers As SyntaxTokenList) As Microsoft.CodeAnalysis.VisualBasic.Syntax.TypeStatementSyntax
			Return Me.WithModifiersCore(modifiers)
		End Function

		Friend MustOverride Function WithModifiersCore(ByVal modifiers As SyntaxTokenList) As Microsoft.CodeAnalysis.VisualBasic.Syntax.TypeStatementSyntax

		Public Function WithTypeParameterList(ByVal typeParameterList As Microsoft.CodeAnalysis.VisualBasic.Syntax.TypeParameterListSyntax) As Microsoft.CodeAnalysis.VisualBasic.Syntax.TypeStatementSyntax
			Return Me.WithTypeParameterListCore(typeParameterList)
		End Function

		Friend MustOverride Function WithTypeParameterListCore(ByVal typeParameterList As Microsoft.CodeAnalysis.VisualBasic.Syntax.TypeParameterListSyntax) As Microsoft.CodeAnalysis.VisualBasic.Syntax.TypeStatementSyntax
	End Class
End Namespace